x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/InferenceServiceNodeLocalRateLimitCalculator.java

```diff
@@ -0,0 +1,231 @@
+        sortedServices.sort(Comparator.comparing(InferenceService::name));
+
+        Map<String, Map<TaskType, RateLimitAssignment>> newAssignments = new HashMap<>();
+
+        for (String serviceName : SERVICE_NODE_LOCAL_RATE_LIMIT_CONFIGS.keySet()) {
+            Optional<InferenceService> service = serviceRegistry.getService(serviceName);
+
+            if (service.isPresent()) {
+                var inferenceService = service.get();
+
+                for (NodeLocalRateLimitConfig rateLimitConfig : SERVICE_NODE_LOCAL_RATE_LIMIT_CONFIGS.get(serviceName)) {
+                    Map<TaskType, RateLimitAssignment> perTaskTypeAssignments = new HashMap<>();
+                    TaskType taskType = rateLimitConfig.taskType();
+
+                    // Calculate node assignments needed for re-routing
+                    var assignedNodes = calculateServiceAssignment(rateLimitConfig.maxNodesPerGroupingStrategy(), sortedNodes);
+
+                    // Update rate limits to be "node-local"
+                    var numAssignedNodes = assignedNodes.size();
+                    var updatedLimitsPerEndpoint = updateRateLimits(inferenceService, numAssignedNodes);
+
+                    perTaskTypeAssignments.put(taskType, new RateLimitAssignment(assignedNodes, updatedLimitsPerEndpoint));
+                    newAssignments.put(serviceName, perTaskTypeAssignments);
+                }
+            } else {
+                logger.warn(
+                    "Service [{}] is configured for node-local rate limiting but was not found in the service registry",
+                    serviceName
+                );
+            }
+        }
+
+        serviceAssignments.set(newAssignments);
+    }
+
+    private List<DiscoveryNode> calculateServiceAssignment(
+        MaxNodesPerGroupingStrategy maxNodesPerGroupingStrategy,
+        List<DiscoveryNode> sortedNodes
+    ) {
+        int numberOfNodes = sortedNodes.size();
+        int nodesPerGrouping = Math.min(numberOfNodes, maxNodesPerGroupingStrategy.calculate(numberOfNodes));
+
+        List<DiscoveryNode> assignedNodes = new ArrayList<>();
+
+        // TODO: here we can probably be smarter: if |num nodes in cluster| > |num nodes per task types|
+        // -> make sure a service provider is not assigned the same nodes for all task types; only relevant as soon as we support more task
+        // types
+        for (int j = 0; j < nodesPerGrouping; j++) {
+            var assignedNode = sortedNodes.get(j % numberOfNodes);
+            assignedNodes.add(assignedNode);
+        }
+
+        return assignedNodes;
+    }
+
+    private Map<String, Long> updateRateLimits(InferenceService service, int responsibleNodes) {
+        if ((service instanceof SenderService) == false) {
+            return Map.of();
+        }
+
+        SenderService senderService = (SenderService) service;
+
+        Sender sender = senderService.getSender();
+        if ((sender instanceof HttpRequestSender) == false) {
```

Comment:
We can avoid the `instanceof` check by adding a new method to the `Sender` interface.

Reply:
Addressed with https://github.com/timgrein/elasticsearch/pull/1